xwiki-platform-core/xwiki-platform-office/xwiki-platform-office-importer/src/main/java/org/xwiki/officeimporter/converter/OfficeConverterResult.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.officeimporter.converter;
+
+import java.io.File;
+import java.util.Set;
+
+import org.xwiki.stability.Unstable;
+
+/**
+ * Provide the result of an office file conversion.
+ * This interface mainly returns the paths of the various files created during the conversion, but it also allows to
+ * cleanup all files created.
+ *
+ * @version $Id$
+ * @since 13.1RC1
+ */
+@Unstable
+public interface OfficeConverterResult
+{
+    /**
+     * @return the output file of the conversion.
+     */
+    File getOutputFile();
+
+    /**
+     * @return the directory containing all output files.
+     */
+    File getOutputDirectory();
+
+    /**
+     * @return all files created during the conversion, including {@link #getOutputFile()}.
+     */
+    Set<File> getAllFiles();
+
+    /**
+     * Allows to completely delete {@link #getOutputDirectory()}.
+     */
+    void cleanup();
```

Comment:
Instead of a custom cleanup() I think I would make OfficeConverterResult Closeable to make clear that this is something that need to be released when you don't need it. It also allows using try/resource and make possible for tools like SonarLint to catch leaks caused by missing close() calls.

Reply:
Hmmm I like the idea but it could be a bit complicated with the current architecture, because of OfficeDocument#getArtifactFiles. This API is actually the one used in script service and those files are directly the ones coming from OfficeConverterResult. So if I perform the cleanup too soon, then the files from OfficeDocument#getArtifactFiles won't exist either which is not right. That's why I added a new API OfficeDocument#cleanupArtifacts which calls the OfficeConverterResult#cleanup, and the former method is called pretty late in the script. 

I'm not a huge fan of this overly complicated archi, but not quite sure how to simplify it since we need: 
  1. a OfficeConverterResult providing info when OfficeConverter#converDocument is called
  2. an OfficeDocument which is returned in script service and should provide all the necessary information

I could make both closeable I guess but then I'd probably need to perform a file copy when building the OfficeDocument, it sounded a bit expensive to me. If you have other ideas to simplify it, please tell me :)

Reply:
Well in my first message I was only suggesting replacing whatever cleanup() does by a call to Closeable#close().

Reply:
Oh I thought you were suggesting to only use OfficeConverterResult in try with resources.

Reply:
But yeah my reaction shows that I don't like really this current archi anyway :)

Reply:
It would definitely be nice if it was possible to release every resource exposed by OfficeConverterResult with a call to close().

Reply:
So I actually implemented it in both OfficeConverterResult and OfficeDocument: the latter relies on the first one, so developers will have to be careful when to call the close. But normally if you're relying on the existing API you're safe.